File: go-sdk/services/oauth_test.go
package services

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"reflect"
	"testing"

	"github.com/pvium/sdks/go-sdk/config"
	"github.com/pvium/sdks/go-sdk/models"
	"github.com/pvium/sdks/go-sdk/transport"
)

type capturedOAuthRequest struct {
	Path     string
	Method   string
	APIKey   string
	Body     map[string]any
	Requests int
}

func createMockOAuthService(t *testing.T) (*OAuthService, *capturedOAuthRequest, func()) {
	t.Helper()

	captured := &capturedOAuthRequest{}
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		captured.Requests++
		captured.Path = r.URL.Path
		captured.Method = r.Method
		captured.APIKey = r.Header.Get("x-api-key")
		if err := json.NewDecoder(r.Body).Decode(&captured.Body); err != nil {
			t.Fatalf("decode request body: %v", err)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"meta":{"statusCode":200,"success":true},"data":{"accessToken":"access_token","refreshToken":"refresh_token","expiresIn":3600}}`))
	}))

	client := transport.NewHTTPClient(config.Config{
		BaseURL:  ts.URL,
		APIKey:   "pk_test_dummy",
		ClientID: "app_test",
	})

	return NewOAuthService(client), captured, ts.Close
}

func TestExchangeCodeForTokenSendsAPIKeyInTokenRequestBody(t *testing.T) {
	t.Parallel()

	service, captured, cleanup := createMockOAuthService(t)
	defer cleanup()

	_, err := service.ExchangeCodeForToken(context.Background(), models.ExchangeAuthorizationCodeInput{
		Code:        "oauth_code",
		RedirectURI: "https://example.test/callback",
	}, nil)
	if err != nil {
		t.Fatalf("exchange code for token: %v", err)
	}

	if captured.Path != "/client-apps/oauth2/token" {
		t.Fatalf("url path mismatch: got %s want %s", captured.Path, "/client-apps/oauth2/token")
	}
	if captured.Method != http.MethodPost {
		t.Fatalf("method mismatch: got %s want %s", captured.Method, http.MethodPost)
	}
	if captured.APIKey != "" {
		t.Fatalf("expected x-api-key header to be omitted, got %q", captured.APIKey)
	}

	wantBody := map[string]any{
		"clientId":    "app_test",
		"apiKey":      "pk_test_dummy",
		"grantType":   "authorization_code",
		"code":        "oauth_code",
		"redirectUri": "https://example.test/callback",
	}
	if !reflect.DeepEqual(captured.Body, wantBody) {
		t.Fatalf("body mismatch:\ngot  %#v\nwant %#v", captured.Body, wantBody)
	}
}

func TestRefreshAccessTokenSendsAPIKeyInTokenRequestBody(t *testing.T) {
	t.Parallel()

	service, captured, cleanup := createMockOAuthService(t)
	defer cleanup()

	_, err := service.RefreshAccessToken(context.Background(), models.RefreshAccessTokenInput{
		RefreshToken: "refresh_token",
	}, nil)
	if err != nil {
		t.Fatalf("refresh access token: %v", err)
	}

	if captured.APIKey != "" {
		t.Fatalf("expected x-api-key header to be omitted, got %q", captured.APIKey)
	}

	wantBody := map[string]any{
		"clientId":     "app_test",
		"apiKey":       "pk_test_dummy",
		"grantType":    "refresh_token",
		"refreshToken": "refresh_token",
	}
	if !reflect.DeepEqual(captured.Body, wantBody) {
		t.Fatalf("body mismatch:\ngot  %#v\nwant %#v", captured.Body, wantBody)
	}
}

func TestGetAccessTokenFromRefreshTokenRefreshesThroughOAuthTokenEndpoint(t *testing.T) {
	t.Parallel()

	service, captured, cleanup := createMockOAuthService(t)
	defer cleanup()

	_, err := service.GetAccessTokenFromRefreshToken(context.Background(), models.RefreshAccessTokenInput{
		RefreshToken: "refresh_token",
	}, nil)
	if err != nil {
		t.Fatalf("get access token from refresh token: %v", err)
	}

	if captured.Path != "/client-apps/oauth2/token" {
		t.Fatalf("url path mismatch: got %s want %s", captured.Path, "/client-apps/oauth2/token")
	}
	if captured.Method != http.MethodPost {
		t.Fatalf("method mismatch: got %s want %s", captured.Method, http.MethodPost)
	}
	if captured.APIKey != "" {
		t.Fatalf("expected x-api-key header to be omitted, got %q", captured.APIKey)
	}

	wantBody := map[string]any{
		"clientId":     "app_test",
		"apiKey":       "pk_test_dummy",
		"grantType":    "refresh_token",
		"refreshToken": "refresh_token",
	}
	if !reflect.DeepEqual(captured.Body, wantBody) {
		t.Fatalf("body mismatch:\ngot  %#v\nwant %#v", captured.Body, wantBody)
	}
}

func TestExchangeCodeForTokenRequiresClientIDAndAPIKey(t *testing.T) {
	t.Parallel()

	client := transport.NewHTTPClient(config.Config{BaseURL: "https://example.com"})
	service := NewOAuthService(client)

	_, err := service.ExchangeCodeForToken(context.Background(), models.ExchangeAuthorizationCodeInput{
		Code:        "code-1",
		RedirectURI: "https://example.com/callback",
	}, nil)
	if err == nil {
		t.Fatal("expected error when clientId and apiKey are missing")
	}
}

File: node-sdk/test/oauth.test.js
const test = require("node:test");
const assert = require("node:assert/strict");

const { PviumApiError, PviumSdk } = require("../dist/index.js");

function createMockSdk(config = {}) {
  const requests = [];
  const sdk = PviumSdk.init({
    baseUrl: "https://api.example.test/v1",
    apiKey: "pk_test_dummy",
    clientId: "app_test",
    fetchFn: async (url, init) => {
      requests.push({ url, init });
      return new Response(
        JSON.stringify({
          meta: { statusCode: 200, success: true },
          data: {
            accessToken: "access_token",
            refreshToken: "refresh_token",
            expiresIn: 3600,
          },
        }),
        { headers: { "content-type": "application/json" } },
      );
    },
    ...config,
  });

  return { sdk, requests };
}

test("exchangeCodeForToken sends apiKey in the token request body", async () => {
  const { sdk, requests } = createMockSdk();

  await sdk.oauth.exchangeCodeForToken({
    code: "oauth_code",
    redirectUri: "https://example.test/callback",
  });

  assert.equal(requests[0].url, "https://api.example.test/v1/client-apps/oauth2/token");
  assert.equal(requests[0].init.method, "POST");
  assert.equal(requests[0].init.headers["x-api-key"], undefined);
  assert.deepEqual(JSON.parse(requests[0].init.body), {
    clientId: "app_test",
    apiKey: "pk_test_dummy",
    grantType: "authorization_code",
    code: "oauth_code",
    redirectUri: "https://example.test/callback",
  });
});

test("refreshAccessToken sends apiKey in the token request body", async () => {
  const { sdk, requests } = createMockSdk();

  await sdk.oauth.refreshAccessToken({
    refreshToken: "refresh_token",
  });

  assert.equal(requests[0].init.headers["x-api-key"], undefined);
  assert.deepEqual(JSON.parse(requests[0].init.body), {
    clientId: "app_test",
    apiKey: "pk_test_dummy",
    grantType: "refresh_token",
    refreshToken: "refresh_token",
  });
});

test("getAccessTokenFromRefreshToken refreshes through the OAuth token endpoint", async () => {
  const { sdk, requests } = createMockSdk();

  await sdk.oauth.getAccessTokenFromRefreshToken({
    refreshToken: "refresh_token",
  });

  assert.equal(requests[0].url, "https://api.example.test/v1/client-apps/oauth2/token");
  assert.equal(requests[0].init.method, "POST");
  assert.equal(requests[0].init.headers["x-api-key"], undefined);
  assert.deepEqual(JSON.parse(requests[0].init.body), {
    clientId: "app_test",
    apiKey: "pk_test_dummy",
    grantType: "refresh_token",
    refreshToken: "refresh_token",
  });
});

test("refreshAccessToken rejects non-2xx responses with PviumApiError", async () => {
  const { sdk } = createMockSdk({
    fetchFn: async () =>
      new Response(
        JSON.stringify({
          meta: {
            statusCode: 401,
            success: false,
            message: "Unauthorized",
            developerMessage: "Refresh token expired",
          },
        }),
        {
          status: 401,
          statusText: "Unauthorized",
          headers: { "content-type": "application/json" },
        },
      ),
  });

  await assert.rejects(
    () => sdk.oauth.refreshAccessToken({ refreshToken: "expired_refresh" }),
    (error) => {
      assert.ok(error instanceof PviumApiError);
      assert.equal(error.status, 401);
      assert.equal(error.message, "Pvium API request failed with 401: Refresh token expired");
      assert.deepEqual(error.body, {
        meta: {
          statusCode: 401,
          success: false,
          message: "Unauthorized",
          developerMessage: "Refresh token expired",
        },
      });
      return true;
    },
  );
});
